### PR TITLE
Allow TextLog body to display as immutable TextEdit rather than Label

### PR DESCRIPTION
**Pull request by tot0**
### What
Convert TextLog body to display as immutable TextEdit rather than Label.
This allows selecting and copying text from the log body.

Used discussion in https://github.com/emilk/egui/discussions/353#discussioncomment-7020619 as inspiration.

`selectable == false` (default):
image

`selectable == true`:
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4716/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4716/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4716/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4716)
- [Docs preview](https://rerun.io/preview/937db0e5dd98a082dd2656cd237417376f737803/docs) 
- [Examples preview](https://rerun.io/preview/937db0e5dd98a082dd2656cd237417376f737803/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

**Comment by tot0** [comment]
Web testing to come...

Appreciate a first look at the text color setting logic since I'm clearly missing something to make it render same as RichText Label does.

**Comment by emilk** [comment]
Thanks for the PR!

We need to support easily copying text from the Rerun UI, but I'm not sure this is the best approach.
It can be very finicky to select the whole text of a label (easy to miss the first letter, or the last letters if it doesn’t fit on screen etc).
It also doesn’t work great on mobile.

A different approach is to use a 📋-button for copying the entire text, like we already do for code-blocks in markdown.
This takes up space, and should probably only show on hover.
For mobile we could have long-press to copy (or long-press to show a context menu, including “copy text”).

I'm gonna talk it through with our designer next week and get back to you!

**Review by emilk — COMMENTED** [review]
After some discussion, we've decided that the drag-to-select approach is indeed the way to go.

We're planning on implementing this wide in egui at some point (https://github.com/emilk/egui/issues/3804), but in the meantime this PR is a good bandaid.

I don't think we should have a "selectable" option - let's just make selectable the default, and make it a good default.

Right now turning on "selectable" also turns on text elision (shortening with "…") which is the opposite of what.

It would also be nice to encapsulate the added code in a `fn selectable_label(ui: &mut egui::Ui, text: egui::RichText)`

**Comment by tot0** [comment]
Thanks @emilk!

I should have some time this weekend to take a second look at this and make selectable the default. There's some aspects I'm missing related to the difference between the RichText label and the immutable TextEdit (masquerading as a Label):
- Colouring changes from grey "inactive" to white "active"
- Column width reducing to fit on screen, rather than expand to as long as the longest body content "off-screen"

Any pointers you might have for where to look to understand the differences are appreciated, otherwise I'll see what I can put together based on the `fn selectable_label(ui: &mut egui::Ui, text: egui::RichText)` abstraction you're proposing.

File: crates/re_space_view_text_log/src/space_view_class.rs [hunk 1]
```diff
@@ -402,16 +406,60 @@ fn table_ui(
...
+                            egui::FontSelection::default()
+                        }
+                        .resolve(ui.style());
+                        let text_color = if let Some(color) = entry.color {
+                            color.into()
+                        } else {
+                            // TODO(lupickup): Understand why TextEdit text color is white instead of gray like the RichText version.
+                            ui.visuals()
+                                .override_text_color
+                                .unwrap_or_else(|| ui.visuals().widgets.inactive.text_color())
```
Comment (emilk) [inline crates/re_space_view_text_log/src/space_view_class.rs:430]: `inactive.text_color` is the text color of an interactive widget (that is not currently being interacted with). That's why you get white text: https://docs.rs/egui/latest/egui/style/struct.Widgets.html#structfield.inactive

Use `noninteractive` instead (we don't count text selection as interaction)

File: crates/re_space_view_text_log/src/space_view_class.rs [hunk 2]
```diff
@@ -402,16 +406,60 @@ fn table_ui(
...
+                            let mut layout_job = egui::text::LayoutJob::simple(
+                                string.to_owned(),
+                                font_id.clone(),
+                                text_color,
+                                wrap_width,
+                            );
+                            layout_job.wrap.max_rows = entry_text_rows;
+                            // Fill to end of body column instead of choosing a full "word" to ellide
+                            layout_job.wrap.break_anywhere = true;
+                            ui.fonts(|f| f.layout_job(layout_job))
```
Comment (emilk) [inline crates/re_space_view_text_log/src/space_view_class.rs:442]: I suggest you compute this `Galley` once and copy into this `layouter`. The `string` you know (and is fixed), and `wrap_width` should be `f32::INFINITY` (no wrapping).

By computing the galley up-front you can use the size of it to size the `TextEdit`.

**Comment by emilk** [comment]
Text selection of all labels will be coming in the next egui release (1-3 weeks away): https://github.com/emilk/egui/pull/3814

**Pull request closed without merging**